[Crash] Duplication of host entries from get hosts query in the database

When XUD (Say Peer1) has previously connected to another XUD (Peer2). Peer2 has Peer1's ip addresses in the hosts table. so during the second time connection between both the peers the hosts data gets exchanged and since peer2 sends the same hosts data to peer1(peer1 table already has it) database is throwing a not unique (`SequelizeUniqueConstraintError`) error.

detailed log:

```bash
> xud@1.0.0-alpha.1 start /home/balamurali_pandranki/xud
> node dist/Xud.js
2018-7-30 19:25:46 [GLOBAL] info: config loaded
Executing (default): SELECT 1+1 AS result
2018-7-30 19:25:47 [DB] info: connected to database. host:******** port:3306 database:xud
2018-7-30 19:25:47 [DB] verbose: Executing (default): CREATE TABLE IF NOT EXISTS `hosts` (`id` INTEGER auto_increment , `address` VARCHAR(39) NOT NULL, `port` INTEGER NOT NULL, `pubKey` VARCHAR(255) UNIQUE, `created
At` DATETIME NOT NULL, `updatedAt` DATETIME NOT NULL, PRIMARY KEY (`id`)) ENGINE=InnoDB;
2018-7-30 19:25:47 [DB] verbose: Executing (default): CREATE TABLE IF NOT EXISTS `bannedHosts` (`address` VARCHAR(39) , `createdAt` DATETIME NOT NULL, `updatedAt` DATETIME NOT NULL, PRIMARY KEY (`address`)) ENGINE=I
nnoDB;
2018-7-30 19:25:47 [DB] verbose: Executing (default): CREATE TABLE IF NOT EXISTS `currencies` (`id` VARCHAR(255) , PRIMARY KEY (`id`)) ENGINE=InnoDB;
2018-7-30 19:25:47 [DB] verbose: Executing (default): SHOW INDEX FROM `hosts`
2018-7-30 19:25:47 [DB] verbose: Executing (default): SHOW INDEX FROM `bannedHosts`
2018-7-30 19:25:47 [DB] verbose: Executing (default): SHOW INDEX FROM `currencies`
2018-7-30 19:25:47 [DB] verbose: Executing (default): CREATE TABLE IF NOT EXISTS `pairs` (`id` VARCHAR(255) , `baseCurrency` VARCHAR(255) NOT NULL, `quoteCurrency` VARCHAR(255) NOT NULL, `swapProtocol` ENUM('LND', '
RAIDEN') NOT NULL, PRIMARY KEY (`id`), FOREIGN KEY (`baseCurrency`) REFERENCES `currencies` (`id`) ON DELETE NO ACTION ON UPDATE CASCADE, FOREIGN KEY (`quoteCurrency`) REFERENCES `currencies` (`id`) ON DELETE NO ACT
ION ON UPDATE CASCADE) ENGINE=InnoDB;
2018-7-30 19:25:47 [DB] verbose: Executing (default): SHOW INDEX FROM `pairs`
2018-7-30 19:25:47 [LND] error: could not find lnd certificate, is lnd installed?
2018-7-30 19:25:47 [LND] info: LndClient status: DISABLED
Executing (default): SELECT `id`, `address`, `port`, `pubKey`, `createdAt`, `updatedAt` FROM `hosts` AS `Host`;
Executing (default): SELECT `address`, `createdAt`, `updatedAt` FROM `bannedHosts` AS `BannedHost`;
Executing (default): SELECT `id`, `baseCurrency`, `quoteCurrency`, `swapProtocol` FROM `pairs` AS `Pair`;
2018-7-30 19:25:47 [RPC] info: gRPC server listening on localhost:8886
2018-7-30 19:25:47 [P2P] info: p2p server listening on 0.0.0.0:8885
2018-7-30 19:28:45 [P2P] debug: Connected to peer (216.38.151.30:64867)
2018-7-30 19:28:45 [P2P] debug: Received data (216.38.151.30:64867): HELLO {"header":{"hash":"CX8h7aZdVS4/daHcHzyvkw=="},"body":{"version":"123","nodePubKey":"123","listenPort":20000,"pairs":["BTC/LTC"]}}
2018-7-30 19:28:45 [P2P] debug: Received data (216.38.151.30:64867): GET_ORDERS {"header":{"hash":"mZFLkyvTelC5g8XnyQrpOw=="},"body":{}}
GET_HOSTS {"header":{"hash":"Vr2O0ewZCBK0RWtFosmEbQ=="},"body":{"ts":1532978925349}}
2018-7-30 19:28:45 [P2P] debug: Received data (216.38.151.30:64867): HOSTS {"header":{"reqHash":"hgWveqDL/OuWkhs/JjpxzA==","hash":"HfKWilOYL2w01WRkMIh3lg=="},"body":{"hosts":[{"instance":{"id":1,"address":"localhost
","port":8886,"pubKey":null,"createdAt":"2018-07-28T06:38:31.000Z","updatedAt":"2018-07-28T06:38:31.000Z"},"socketAddress":{"address":"localhost","port":8886}},{"instance":{"id":2,"address":"35.196.118.79","port":88
85,"pubKey":null,"createdAt":"2018-07-30T18:36:39.000Z","updatedAt":"2018-07-30T18:36:39.000Z"},"socketAddress":{"address":"35.196.118.79","port":8885}}]}}
ORDERS {"header":{"hash":"sLrRmEfx3k9eLlkpWRUyQg=="},"body":{"orders":[]}}
2018-7-30 19:28:45 [P2P] debug: Peer (216.38.151.30:64867) sent an unsolicited response packet (ORDERS)
2018-7-30 19:28:45 [P2P] debug: Connected to peer ([object Object])
2018-7-30 19:28:45 [P2P] debug: Ignoring banned peer (35.196.118.79)
2018-7-30 19:28:45 [P2P] debug: Connected to peer ([object Object])
2018-7-30 19:28:45 [P2P] debug: Socket Error ([object Object]): Error: read ECONNRESET
2018-7-30 19:28:45 [P2P] error: peer error Error: read ECONNRESET
    at _errnoException (util.js:992:11)
    at TCP.onread (net.js:618:25)
2018-7-30 19:28:45 [P2P] warn: disconnected peer ([object Object]) hostId is missing
2018-7-30 19:28:45 [P2P] debug: Received data ([object Object]): ?@@ �?
2018-7-30 19:28:45 [P2P] warn: Unknown peer message type: ?@@
2018-7-30 19:28:45 [P2P] debug: Socket Error ([object Object]): Socket hangup
2018-7-30 19:28:45 [P2P] error: peer error Error: Socket Error ([object Object]): Socket hangup
    at Peer.error (/home/balamurali_pandranki/xud/dist/p2p/Peer.js:333:36)
    at Socket.socket.once (/home/balamurali_pandranki/xud/dist/p2p/Peer.js:277:22)
    at Object.onceWrapper (events.js:315:30)
    at emitOne (events.js:116:13)
    at Socket.emit (events.js:211:7)
    at TCP._handle.close [as _onclose] (net.js:557:12)
2018-7-30 19:28:45 [P2P] warn: disconnected peer ([object Object]) hostId is missing
Executing (default): SELECT `id`, `baseCurrency`, `quoteCurrency`, `swapProtocol` FROM `pairs` AS `Pair`;
Executing (default): INSERT INTO `hosts` (`id`,`address`,`port`,`createdAt`,`updatedAt`) VALUES (DEFAULT,'35.196.118.79',8885,'2018-07-30 19:28:45','2018-07-30 19:28:45');
Executing (default): INSERT INTO `hosts` (`id`,`address`,`port`,`createdAt`,`updatedAt`) VALUES (DEFAULT,'localhost',8886,'2018-07-30 19:28:45','2018-07-30 19:28:45');
{ SequelizeUniqueConstraintError: Validation error
    at Query.formatError (/home/balamurali_pandranki/xud/node_modules/sequelize/lib/dialects/mysql/query.js:226:16)
    at Query.handler [as onResult] (/home/balamurali_pandranki/xud/node_modules/sequelize/lib/dialects/mysql/query.js:68:23)
    at Query.Command.execute (/home/balamurali_pandranki/xud/node_modules/mysql2/lib/commands/command.js:30:12)
    at Connection.handlePacket (/home/balamurali_pandranki/xud/node_modules/mysql2/lib/connection.js:502:28)
    at PacketParser.onPacket (/home/balamurali_pandranki/xud/node_modules/mysql2/lib/connection.js:81:16)
    at PacketParser.executeStart (/home/balamurali_pandranki/xud/node_modules/mysql2/lib/packet_parser.js:77:14)
    at Socket.<anonymous> (/home/balamurali_pandranki/xud/node_modules/mysql2/lib/connection.js:89:29)
    at emitOne (events.js:116:13)
    at Socket.emit (events.js:211:7)
    at addChunk (_stream_readable.js:263:12)
    at readableAddChunk (_stream_readable.js:250:11)
    at Socket.Readable.push (_stream_readable.js:208:10)
    at TCP.onread (net.js:597:20)
  name: 'SequelizeUniqueConstraintError',
  errors: 
   [ ValidationErrorItem {
       message: 'hosts_address_port must be unique',
       type: 'unique violation',
       path: 'hosts_address_port',
       value: '35.196.118.79-8885',
       origin: 'DB',
       instance: [Object],
       validatorKey: 'not_unique',
       validatorName: null,
       validatorArgs: [] } ],
  fields: { hosts_address_port: '35.196.118.79-8885' },
  parent: 
   { Error: Duplicate entry '35.196.118.79-8885' for key 'hosts_address_port'
    at Packet.asError (/home/balamurali_pandranki/xud/node_modules/mysql2/lib/packets/packet.js:716:13)
    at Query.Command.execute (/home/balamurali_pandranki/xud/node_modules/mysql2/lib/commands/command.js:28:22)
    at Connection.handlePacket (/home/balamurali_pandranki/xud/node_modules/mysql2/lib/connection.js:502:28)
    at PacketParser.onPacket (/home/balamurali_pandranki/xud/node_modules/mysql2/lib/connection.js:81:16)
    at PacketParser.executeStart (/home/balamurali_pandranki/xud/node_modules/mysql2/lib/packet_parser.js:77:14)
    at Socket.<anonymous> (/home/balamurali_pandranki/xud/node_modules/mysql2/lib/connection.js:89:29)
    at emitOne (events.js:116:13)
    at Socket.emit (events.js:211:7)
    at addChunk (_stream_readable.js:263:12)
    at readableAddChunk (_stream_readable.js:250:11)
    at Socket.Readable.push (_stream_readable.js:208:10)
    at TCP.onread (net.js:597:20)
     code: 'ER_DUP_ENTRY',
     errno: 1062,
     sqlState: '23000',
     sqlMessage: 'Duplicate entry \'35.196.118.79-8885\' for key \'hosts_address_port\'',
     sql: 'INSERT INTO `hosts` (`id`,`address`,`port`,`createdAt`,`updatedAt`) VALUES (DEFAULT,\'35.196.118.79\',8885,\'2018-07-30 19:28:45\',\'2018-07-30 19:28:45\');' },
  original: 
   { Error: Duplicate entry '35.196.118.79-8885' for key 'hosts_address_port'
    at Packet.asError (/home/balamurali_pandranki/xud/node_modules/mysql2/lib/packets/packet.js:716:13)
    at Query.Command.execute (/home/balamurali_pandranki/xud/node_modules/mysql2/lib/commands/command.js:28:22)
    at Connection.handlePacket (/home/balamurali_pandranki/xud/node_modules/mysql2/lib/connection.js:502:28)
    at PacketParser.onPacket (/home/balamurali_pandranki/xud/node_modules/mysql2/lib/connection.js:81:16)
    at PacketParser.executeStart (/home/balamurali_pandranki/xud/node_modules/mysql2/lib/packet_parser.js:77:14)
    at Socket.<anonymous> (/home/balamurali_pandranki/xud/node_modules/mysql2/lib/connection.js:89:29)
    at emitOne (events.js:116:13)
    at Socket.emit (events.js:211:7)
    at addChunk (_stream_readable.js:263:12)
    at readableAddChunk (_stream_readable.js:250:11)
    at Socket.Readable.push (_stream_readable.js:208:10)
    at TCP.onread (net.js:597:20)
     code: 'ER_DUP_ENTRY',
     errno: 1062,
     sqlState: '23000',
     sqlMessage: 'Duplicate entry \'35.196.118.79-8885\' for key \'hosts_address_port\'',
     sql: 'INSERT INTO `hosts` (`id`,`address`,`port`,`createdAt`,`updatedAt`) VALUES (DEFAULT,\'35.196.118.79\',8885,\'2018-07-30 19:28:45\',\'2018-07-30 19:28:45\');' },
  sql: 'INSERT INTO `hosts` (`id`,`address`,`port`,`createdAt`,`updatedAt`) VALUES (DEFAULT,\'35.196.118.79\',8885,\'2018-07-30 19:28:45\',\'2018-07-30 19:28:45\');' }
/home/balamurali_pandranki/xud/dist/bootstrap.js:13
        throw err;
        ^

SequelizeUniqueConstraintError: Validation error
    at Query.formatError (/home/balamurali_pandranki/xud/node_modules/sequelize/lib/dialects/mysql/query.js:226:16)
    at Query.handler [as onResult] (/home/balamurali_pandranki/xud/node_modules/sequelize/lib/dialects/mysql/query.js:68:23)
    at Query.Command.execute (/home/balamurali_pandranki/xud/node_modules/mysql2/lib/commands/command.js:30:12)
    at Connection.handlePacket (/home/balamurali_pandranki/xud/node_modules/mysql2/lib/connection.js:502:28)
    at PacketParser.onPacket (/home/balamurali_pandranki/xud/node_modules/mysql2/lib/connection.js:81:16)
    at PacketParser.executeStart (/home/balamurali_pandranki/xud/node_modules/mysql2/lib/packet_parser.js:77:14)
    at Socket.<anonymous> (/home/balamurali_pandranki/xud/node_modules/mysql2/lib/connection.js:89:29)
    at emitOne (events.js:116:13)
    at Socket.emit (events.js:211:7)
    at addChunk (_stream_readable.js:263:12)
    at readableAddChunk (_stream_readable.js:250:11)
    at Socket.Readable.push (_stream_readable.js:208:10)
    at TCP.onread (net.js:597:20)
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! xud@1.0.0-alpha.1 start: `node dist/Xud.js`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the xud@1.0.0-alpha.1 start script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
```

--- a/lib/db/models/Host.ts
+++ b/lib/db/models/Host.ts
@@ -6,7 +6,7 @@ export default (sequelize: Sequelize.Sequelize, DataTypes: Sequelize.DataTypes) 
     id: { type: DataTypes.INTEGER, primaryKey: true, autoIncrement: true },
     address: { type: DataTypes.STRING(39), allowNull: false },
     port: { type: DataTypes.INTEGER, allowNull: false },
-    pubKey: { type: DataTypes.STRING, allowNull: true, unique: true },
+    pubKey: { type: DataTypes.STRING, allowNull: true, unique: false },
   };
 
   const indexes: Sequelize.DefineIndexesOptions[] = [{
@@ -19,7 +19,7 @@ export default (sequelize: Sequelize.Sequelize, DataTypes: Sequelize.DataTypes) 
     tableName: 'hosts',
   };
 
-  const Peer = sequelize.define<db.HostInstance, db.HostAttributes>('Host', attributes, options);
+  const Host = sequelize.define<db.HostInstance, db.HostAttributes>('Host', attributes, options);
 
-  return Peer;
+  return Host;
 };

--- a/lib/p2p/HostList.ts
+++ b/lib/p2p/HostList.ts
@@ -2,6 +2,8 @@ import P2PRepository from './P2PRepository';
 import SocketAddress from './SocketAddress';
 import Host from './Host';
 import Peer from './Peer';
+import assert from 'assert';
+import { HostFactory } from '../types/db';
 
 /** Represents a list of hosts for managing network peers activity */
 class HostList {
@@ -24,11 +26,32 @@ class HostList {
   }
 
   /**
-   * Get a host by socket address, or create the host if none exists.
+   * Return an array of hosts that haven't been banned.
    */
-  public getOrCreateHost = async (socketAddress: SocketAddress): Promise<Host> => {
+  public toHostFactoryArray = (): HostFactory[] => {
+    const hostValues = Object.values(this.hosts);
+    const hostFactories: HostFactory[] = Array.from({ length: hostValues.length });
+    for (let i = 0; i < hostValues.length; i += 1) {
+      const hostValue = hostValues[i];
+      hostFactories[i] = {
+        address: hostValue.socketAddress.address,
+        port: hostValue.socketAddress.port,
+      };
+    }
+    return hostFactories;
+  }
+
+  /**
+   * Get a host by a peer's socket address, or create the host if none exists.
+   */
+  public getOrCreateHost = async (peer: Peer): Promise<Host> => {
+    const { socketAddress } = peer;
+    assert(socketAddress instanceof SocketAddress);
     const host = this.hosts[socketAddress.toString()];
-    return host || this.createHost(socketAddress);
+    return host || this.createHost({
+      address: socketAddress.address,
+      port: socketAddress.port,
+    });
   }
 
   public ban = async (peer: Peer): Promise<void> => {
@@ -55,8 +78,8 @@ class HostList {
     this.bannedHosts = new Set<string>(bannedHosts.map(bannedHost => bannedHost.address));
   }
 
-  private createHost = async (socketAddress: SocketAddress): Promise<Host> => {
-    const dbHost = await this.repository.addHost(socketAddress);
+  private createHost = async (hostFactory: HostFactory): Promise<Host> => {
+    const dbHost = await this.repository.addHost(hostFactory);
     const host = new Host(dbHost);
     this.hosts[host.socketAddress.toString()] = host;
     return host;

--- a/lib/p2p/Peer.ts
+++ b/lib/p2p/Peer.ts
@@ -9,6 +9,7 @@ import Logger from '../Logger';
 import { ms } from '../utils/utils';
 import { OutgoingOrder } from '../types/orders';
 import { Packet, PacketDirection, PacketType } from './packets';
+import { HostFactory } from '../types/db';
 
 enum ConnectionDirection {
   INBOUND,
@@ -190,7 +191,7 @@ class Peer extends EventEmitter {
     this.sendPacket(packet);
   }
 
-  public sendHosts = (hosts: Host[], reqId: string): void => {
+  public sendHosts = (hosts: HostFactory[], reqId: string): void => {
     const packet = new packets.HostsPacket(hosts, reqId);
     this.sendPacket(packet);
   }

--- a/lib/p2p/Pool.ts
+++ b/lib/p2p/Pool.ts
@@ -180,14 +180,14 @@ class Pool extends EventEmitter {
         break;
       }
       case PacketType.GET_HOSTS: {
-        peer.sendHosts(this.hosts.toArray(), packet.header.id);
+        peer.sendHosts(this.hosts.toHostFactoryArray(), packet.header.id);
         break;
       }
       case PacketType.HOSTS: {
         const hosts = (packet as HostsPacket).body!;
         hosts.forEach(async (host) => {
           try {
-            await this.addOutbound(host.socketAddress);
+            await this.addOutbound(new SocketAddress(host.address, host.port));
           } catch (err) {
             this.logger.info(err);
           }
@@ -207,11 +207,11 @@ class Pool extends EventEmitter {
 
   private setPeerHost = async (peer: Peer, listenPort?: number): Promise<void> => {
     if (peer.direction === ConnectionDirection.OUTBOUND) {
-      const host = await this.hosts.getOrCreateHost(peer.socketAddress);
+      const host = await this.hosts.getOrCreateHost(peer);
       peer.setHost(host);
     } else if (peer.direction === ConnectionDirection.INBOUND && listenPort) {
       const socketAddress = new SocketAddress(peer.socketAddress.address, listenPort);
-      const host = await this.hosts.getOrCreateHost(socketAddress);
+      const host = await this.hosts.getOrCreateHost(peer);
       peer.setHost(host);
     }
   }

--- a/lib/p2p/packets/types/HostsPacket.ts
+++ b/lib/p2p/packets/types/HostsPacket.ts
@@ -1,10 +1,8 @@
 import Packet, { PacketDirection } from '../Packet';
 import PacketType from '../PacketType';
-import Host from '../../Host';
+import { HostFactory } from '../../../types/db';
 
-type HostsPacketBody = Host[];
-
-class HostsPacket extends Packet<HostsPacketBody> {
+class HostsPacket extends Packet<HostFactory[]> {
   public get type() {
     return PacketType.HOSTS;
   }

